Add RabbitMQ backend for event_pusher

mkacper opened the pull request.

mongoose-im commented:
[5836.3](https://travis-ci.org/esl/MongooseIM/jobs/452415951) / Erlang 19.3 / mysql_redis / 61da16f4b9c0642cf2505d18973dff8f71664ecf
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2127/5836/mysql_redis.19.3)/ [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2127/5836/mysql_redis.19.3/big/ct_run.test%40travis-job-951b7035-e0ef-4c7a-a8eb-3bb975b31845.2018-11-08_15.36.16/index.html)
**OK: 2888** / Failed: 0 / **User-skipped: 222** / Auto-skipped: 0

mongoose-im commented:
[5838.1](https://travis-ci.org/esl/MongooseIM/jobs/452776500) / Erlang 19.3 / small_tests / bbf7c58a4a601e79343e65a9b62daa6b3a079aa8
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2127/5838/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2127/5838/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-09_09.13.26/lib.mongooseim.logs/run.2018-11-09_09.13.27/suite.log.html)

mkacper commented on .travis.yml:
```diff
@@ -33,6 +33,7 @@ install:
 before_script:
         - tools/travis-setup-db.sh
         - if [ $PRESET = 'ldap_mnesia' ]; then sudo tools/travis-setup-ldap.sh; fi
+        - if [ $PRESET = 'mysql_redis' ]; then sudo tools/travis-setup-rmq.sh; fi
```
Because this suite is fairly short. We've discussed this with @michalwski.

mkacper commented on doc/modules/mod_event_pusher_rabbit.md:
```diff
@@ -0,0 +1,193 @@
+The module provides some metrics related to RabbitMQ connections and messages
+as well. Provided metrics:
+
+  * `connections_active` - number of active connections to a RabbitMQ
```
That's right. What do you suggest then? IMO it'd be enough to put a notice into this part of the documentation. What do you reckon?

mkacper commented on doc/modules/mod_event_pusher_rabbit.md:
```diff
@@ -0,0 +1,193 @@
+section). When a worker sends a message to a RabbitMQ server it waits for a
+confirmation from the server before it starts to process next message. This
+approach allows to introduce backpressure on a RabbitMQ server connection cause
+the server can reject messages when it's overloaded. On the other hand it can
```
I don't know. `amqp_channel:wait_for_confirms/1` can return `true`, `false` or `timeout`.

mkacper commented on doc/modules/mod_event_pusher_rabbit.md:
```diff
@@ -0,0 +1,193 @@
+
+#### Event messages ordering
+
+`available_worker` strategy does not ensure that user events will be delivered to
```
Because we use `cast` in `mod_event_pusher_rabbit` so I can imagine that the worker which received the first message is processing it longer that the one which received the second message. The reason for that might be the first message is simply larger that the second one. Does it make sense?

telezynski replied:
Could be if there are multiple workers used - there is a race between workers on delivering messages.
Strict ordering requires serialization of message delivery by employing one worker only.
So actually mode of queuing doesn't matter (if workers are not restarted, that is if they do not loose mbox), it is only a performance optimization, what matters is the number of configured workers.

arcusfelis commented on doc/advanced-configuration/outgoing-connections.md:
```diff
@@ -450,3 +451,44 @@ Run the following function in the MongooseIM shell to verify that the connection
+ {rabbit, host, event_pusher, [{workers, 20}],
+  [{amqp_host, "localhost"},
+   {amqp_port, 5672},
+   {amqp_username, <<"guest">>},
```
I would rather do not use binaries in configs. Using strings for amqp_username and amqp_password  would be more consistent.

mkacper replied:
Good point! However, I can only change that for the connection parameters. Changing this for exchanges would end up with calling `binary:list_to_bin/1` for every "event" coming in.

arcusfelis commented on test/mongoose_rabbit_worker_SUITE.erl:
```diff
@@ -0,0 +1,205 @@
+
+get_worker_conn_and_chann(Worker) ->
+    State = sys:get_state(Worker),
+    {element(3, State), element(4, State)}.
```
oh, this code can break, if someone would edit the worker code.

One way to improve it is to implement the callback: 
http://erlang.org/doc/man/gen_server.html#Module:format_status-2

Which returns a map or a proplist of the record fields from it.
Of course, this callback on it's own would require its own tests :)

mkacper replied:
Agree, that's naive implementation.

I can handle relevant test cases using `is_in_worker_state/2` implemented as follows:

```
is_in_worker_state(Worker, Elem) ->
    State = sys:get_state(Worker),
    lists:member(Elem, tuple_to_list(State)).
```

Does it make much sense now?

erszcz replied:
Looks way better, IMO.

erszcz replied:
Looks way better, IMO.

arcusfelis commented on src/wpool/mongoose_wpool_rabbit.erl:
```diff
@@ -0,0 +1,41 @@
+    ok.
+
+amqp_client_opts(AMQPOpts) ->
+    Opts = [{host, proplists:get_value(amqp_host, AMQPOpts, undefined)},
```
`proplists:get_value(amqp_host, AMQPOpts, undefined)` can  be written as `proplists:get_value(amqp_host, AMQPOpts)`

arcusfelis commented on src/wpool/mongoose_wpool_rabbit.erl:
```diff
@@ -0,0 +1,41 @@
+              [{amqp_client_opts, amqp_client_opts(AMQPOpts)},
+               {host, Host},
+               {confirms, IsConfirmEnabled},
+               {max_queue_len, ?MAX_WORKER_MSG_QUEUE_LEN}]},
```
It should be configurable.

arcusfelis commented on src/mongoose_rabbit_worker.erl:
```diff
@@ -0,0 +1,302 @@
+            {noreply, Opts#state{connection = FreshConn, channel = FreshChann}};
+        timeout ->
+            update_messages_timeout_metrics(Host),
+            ?WARNING_MSG("event=rabbit_message_sent_failed reason=timeout",
```
It's short enough to be on one line:

```erlang
?WARNING_MSG("event=rabbit_message_sent_failed reason=timeout", []),
```

arcusfelis commented on src/mongoose_rabbit_worker.erl:
```diff
@@ -0,0 +1,302 @@
+%%%===================================================================
+
+do_handle_call({amqp_call, Method}, _From, State = #state{channel = Channel}) ->
+    case try_to_apply(amqp_channel, call, [Channel, Method]) of
```
Use `try ... catch` here.
Dialyzer can't handle`erlang:apply/3` calls. `amqp_channel:call(Channel, Method)` should be used instead.

erszcz replied:
I second @arcusfelis's remark. I do not see a reason not to use `amqp_channel:call(Channel, Method)` directly and also to hide the `try ... catch` block in a separate function. The overall structure of code is almost identical, while the cognitive overhead of non-local exception handling is avoided:

```erlang
try
    Res = ampq_channel:call(Channel, Method),
    {reply, Res, State}
catch E:R ->
    {FreshConn, FreshChann} = maybe_restart_rabbit_connection(State),
    {reply, {E, R}, State#state{connection = FreshConn,
                                channel = FreshChann}}
end
```

arcusfelis commented on src/mongoose_rabbit_worker.erl:
```diff
@@ -0,0 +1,302 @@
+-spec restart_rabbit_connection(worker_opts()) -> {pid(), pid()}.
+restart_rabbit_connection(#state{connection = Conn, channel = Chann,
+                                 host = Host, amqp_client_opts = AMQPOpts}) ->
+    catch close_rabbit_connection(Conn, Chann, Host),
```
The code ignores all the errors, the catch would ignore such errors as "Conn = undefined" for example.

arcusfelis commented on src/wpool/mongoose_wpool_rabbit.erl:
```diff
@@ -0,0 +1,41 @@
+    ok.
+
+amqp_client_opts(AMQPOpts) ->
+    Opts = [{host, proplists:get_value(amqp_host, AMQPOpts, undefined)},
```
How do we handle case if our user forgot to set host argument or set it to something strange?
I guess the data goes directly into rabbitmq client without any verification.

mkacper replied:
`mongoose_amqp:network_params/1` sets default values when a user doesn't pass some of the required parameters. But when it passes something strange a worker won't be able to connect to a Rabbit server. Should we validate these config values?

mkacper commented on doc/modules/mod_event_pusher_rabbit.md:
```diff
@@ -0,0 +1,193 @@
+### Module Description
```
It's automatically created (AFAIU by `mkdocs`).

mkacper commented on src/event_pusher/mod_event_pusher_rabbit.erl:
```diff
@@ -0,0 +1,232 @@
+    cast_rabbit_worker(Host, {amqp_publish, PublishMethod, AMQPMessage}).
+
+-spec handle_user_chat_event(#chat_event{}) -> ok.
+handle_user_chat_event(#chat_event{type = normal}) -> ok;
```
I can see this type of messages coming in after merge with newer master branch but we don't handle this kind of messages for now. The other thing is if we should handle them. Maybe someone from Mongoose team can speak up?

mkacper commented on src/wpool/mongoose_wpool_rabbit.erl:
```diff
@@ -0,0 +1,41 @@
+-export([start/4]).
+-export([stop/2]).
+
+%% If a worker message queue length reaches the limit, messages from the head of
```
We drop the oldest message. It's because we drop messages as we're processing them.

mkacper commented on test/mongoose_rabbit_worker_SUITE.erl:
```diff
@@ -0,0 +1,205 @@
+    Config ++ [{worker_pid, WorkerPid}].
+
+end_per_testcase(_Case, Config) ->
+    exit(proplists:get_value(worker_pid, Config), kill),
```
I wanted to stop the process as quickly as possible. Is that wrong practise in tests?

mkacper commented:
@mentels please see my comments.

mongoose-im commented:
[5860.1](https://travis-ci.org/esl/MongooseIM/jobs/455425551) / Erlang 19.3 / small_tests / 29eb928eaa62c3a4a201e2014fc81e165f5b804c
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2127/5860/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2127/5860/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-15_12.25.37/lib.mongooseim.logs/run.2018-11-15_12.25.38/suite.log.html)

erszcz commented on src/mongoose_rabbit_worker.erl:
```diff
@@ -0,0 +1,302 @@
+            ?WARNING_MSG("event=rabbit_message_sent_failed reason=negative_ack",
+                         []),
+            {noreply, Opts};
+        {channel_exception, Error, Reason} ->
```
It seems that `maybe_restart_rabbit_connection` reuses the connection if it's still open.

mkacper replied:
I've improved that recently (after Szymon's remark).

erszcz commented on src/mongoose_rabbit_worker.erl:
```diff
@@ -0,0 +1,306 @@
+publish_message_and_wait_for_confirm(Method, Payload,
+                                     #state{channel = Channel,
+                                            confirms = IsConfirmEnabled}) ->
+    case try_to_apply(amqp_channel, call, [Channel, Method, Payload]) of
```
Same arguments as for `amqp_channel:call/2` above apply here.

erszcz commented on src/mongoose_rabbit_worker.erl:
```diff
@@ -0,0 +1,306 @@
+-spec publish_message_and_wait_for_confirm(Method :: mongoose_amqp:method(),
+                                           Payload :: mongoose_amqp:message(),
+                                           worker_opts()) ->
+    boolean() | timeout | channel_exception.
```
It's a matter of preference, but when a return value is complex (or there are complex parameter types involved) I like to write the specs according to the following pattern - it's more suitable for inserting line breaks for readability:

```erlang
-spec publish_message_and_wait_for_confirm(Method, Payload, worker_opts()) -> R
      when Method :: mongoose_amqp:method(),
           Payload :: mongoose_amqp:message(),
           R :: boolean() | timeout
              %% this line break is a stretch, but shows an example
              | channel_exception.
```

mkacper commented:
@arcusfelis please see my comments. I also added some of the requested changes.

I also changed this "naive" test case a bit. See here: https://github.com/esl/MongooseIM/pull/2127/commits/8494382feed9bff07a3d97cab0b4f0c85953dabe

cc @erszcz

mongoose-im commented:
[5868.1](https://travis-ci.org/esl/MongooseIM/jobs/455951880) / Erlang 19.3 / small_tests / 7f1611355e9190f8e0d6a60ea003c341f45cf83b
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2127/5868/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2127/5868/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-16_12.34.49/lib.mongooseim.logs/run.2018-11-16_12.34.49/suite.log.html)

erszcz commented on doc/modules/mod_event_pusher_rabbit.md:
```diff
@@ -0,0 +1,193 @@
+There are no guarantees. The current implementation uses "best effort" approach
+which means that we don't care if a message is delivered to a RabbitMQ server.
+If a message couldn't be delivered to the server for any reason the module
+just updates appropriate metrics and print some log messages.
```
prints

erszcz commented on doc/modules/mod_event_pusher_rabbit.md:
```diff
@@ -0,0 +1,193 @@
+gives a lot of flexibility when binding queues to such an exchange by using
+`#` and `*` in binding keys. But flexibility comes at the cost of performance -
+imagine a scenario where there are thousands of users and AMQP consumers use
+bindig keys for particular users which looks like `user_N@host.#`. In such
```
> bindig keys for particular users which looks like `user_N@host.#`. 
bind>in<g keys ... look like ...

erszcz commented on doc/modules/mod_event_pusher_rabbit.md:
```diff
@@ -0,0 +1,193 @@
+case RabbitMQ has to go through all the users in order to find out where
+a message should be sent to. This operations is proved to be costly. In a load
+test with 100k users a delay caused by this operation was substantial (about an
+order of magnitue higher than compared to a load test with 60k users).
```
magnitude

erszcz commented on doc/modules/mod_event_pusher_rabbit.md:
```diff
@@ -0,0 +1,193 @@
+confirmation from the server before it starts to process next message. This
+approach allows to introduce backpressure on a RabbitMQ server connection cause
+the server can reject messages when it's overloaded. On the other hand it can
+cause performance degratation.
```
degradation

erszcz commented on src/mongoose_rabbit_worker.erl:
```diff
@@ -0,0 +1,298 @@
+        true ->
+            ConfirmCallRes = mongoose_amqp:confirm_select_ok(),
+            ConfirmCallRes =
+                amqp_channel:call(Channel, mongoose_amqp:confirm_select()),
```
This `amqp_channel:call` is not guarded by a `try ... catch` like the other ones. I'm not sure if it's important in this case.

mkacper replied:
This is one of worker's initialisation steps (enabling message confirmations from a RabbitMQ server). Should the step fail the worker won't be started.

erszcz commented:
@mkacper The last two Travis builds are either cancelled or red and it's a strong warning. Is this still in progress?

mkacper commented:
@erszcz I've been still introducing some changes required in the reviews. The latest build was cancelled by me as it changes only the documentation. The one before the last was cancelled to free up some  resources on Travis for builds related to the latest MongooseIM release.

mongoose-im commented:
[5910.1](https://travis-ci.org/esl/MongooseIM/jobs/459866268) / Erlang 19.3 / small_tests / 1589bb94a245e0df3572aaf11ce10268aae5af48
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2127/5910/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2127/5910/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-11-26_17.57.04/lib.mongooseim.logs/run.2018-11-26_17.57.05/suite.log.html)

mentels commented:
Any plans to merge this?

michalwski commented:
Yes! We want and are going to merge this. Most probably next year though :)

fenek commented on src/event_pusher/mod_event_pusher_rabbit.erl:
```diff
@@ -0,0 +1,231 @@
+    cast_rabbit_worker(Host, {amqp_publish, PublishMethod, AMQPMessage}).
+
+-spec handle_user_chat_event(#chat_event{}) -> ok.
+handle_user_chat_event(#chat_event{type = normal}) -> ok;
```
It's better to match for only `chat` and `groupchat` explicitly and ignore everything else. Right now, this function will crash if after some changes in the future, a `headline` messages reaches this function.

fenek commented on src/mongoose_rabbit_worker.erl:
```diff
@@ -0,0 +1,298 @@
+%%% Metrics
+%%%===================================================================
+
+-define(BACKEND, mod_event_pusher_rabbit).
```
This really shouldn't be hardcoded here, as this module doesn't necessarily have to be used by Event Pusher only.

mkacper replied:
What do you suggest then?

mkacper replied:
`BACKEND` is only used for metrics names so maybe we can just pass wpool tag to the worker (as you mentioned in the next comment) and use it instead?

fenek commented on doc/modules/mod_event_pusher_rabbit.md:
```diff
@@ -0,0 +1,198 @@
+  a RabbitMQ server (including message properties)
+
+> All the above metrics have a prefix which looks as follows:  
+> `mongooseim.<xmpp_host>.backends.mod_event_pusher_rabbit.<metric_name>`.
```
`mongooseim` prefix may be added e.g. in Wombat or graphite but in MongooseIM there is no such metric name element.

fenek commented on src/mongoose_rabbit_worker.erl:
```diff
@@ -0,0 +1,298 @@
+%%%===================================================================
+
+-define(BACKEND, mod_event_pusher_rabbit).
+-define(CONNECTIONS_ACTIVE_METRIC, [backends, ?BACKEND, connections_active]).
```
I think it's better to pass pool `Tag` to the worker and use it in metric name.

fenek commented on src/mongoose_rabbit_worker.erl:
```diff
@@ -0,0 +1,298 @@
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
+         terminate/2]).
+
+-type state() :: #{amqp_client_opts := mongoose_amqp:network_params(),
```
This type definition is most probably incorrect. Try adding spec to `init` function. I'm pretty sure all of these have to be defined with `=>`, not `:=`.

mkacper replied:
Here is what I've found in Erlang docs:

```
Association :: Type := Type                  %% denotes a mandatory association
             | Type => Type                  %% denotes an optional association

```

mkacper replied:
Now I see, you're right :)

The connection is done asynchronously to start workers faster.

fenek replied:
I think we have a race condition then. What if a publish request arrives before the init is complete?

These pools will usually be started on MongooseIM startup, so it's not that important to save like how much? 1s? 2s? Other pool workers are initialised synchronously anyway.

fenek commented on src/mongoose_rabbit_worker.erl:
```diff
@@ -0,0 +1,298 @@
+
+-type worker_opts() :: state().
+
+-define(REPLY_REQ_DROPPED(State), {reply, request_dropped, State}).
```
These macros are used in only 3 places and given their length, there is no gain from introducing them.

fenek commented on src/mongoose_rabbit_worker.erl:
```diff
@@ -0,0 +1,298 @@
+            {noreply, Opts};
+        {channel_exception, Error, Reason} ->
+            update_messages_failed_metrics(Host),
+            ?WARNING_MSG("event=rabbit_message_sent_failed reason=~1000p:~1000p",
```
So few details. No `Host`, no basic connection params that can be used for debugging... This applies to other logs in this file as well.

fenek commented on src/mongoose_rabbit_worker.erl:
```diff
@@ -0,0 +1,298 @@
+        {error, Error} ->
+            update_failed_connections_metrics(Host),
+            ?ERROR_MSG("event=rabbit_connection_failed reason=~1000p", [Error]),
+            exit("connection to a Rabbit server failed")
```
Sooooo few details.

fenek replied:
Still not enough details.

fenek replied:
Nvm. :) Haven't noticed the line above.

fenek commented on src/mongoose_rabbit_worker.erl:
```diff
@@ -0,0 +1,298 @@
+            apply(Callback, Args);
+        true ->
+            ?WARNING_MSG("event=rabbit_worker_request_dropped "
+                         ++ "reason=queue_message_length_limit_reached "
```
There is no need to add `++` here. Strings will be catenated anyway.

fenek commented on src/mongoose_rabbit_worker.erl:
```diff
@@ -0,0 +1,298 @@
+    gen_server:start_link(?MODULE, [], []).
+
+list_metrics() ->
+    [{?CONNECTIONS_ACTIVE_METRIC, spiral},
```
This one should be `value` or `gauge`.

mkacper replied:
Why `spiral` is wrong here? We just increment the metric when a connection is opened and decrement when it's closed. If we use e.g. `gauge` we have to calculate somehow a total number of active connections. It seems that using `spiral` here is fairly simple solution.

fenek replied:
OK, right, I remembered metric type improperly. :( I had `counter` in mind. `spiral` is not *wrong* as using it's `count` property indeed works. But its feature to track value changes in 1-minute window is kind of useless here and `count` makes more sense semantically.

fenek commented on doc/modules/mod_event_pusher_rabbit.md:
```diff
@@ -0,0 +1,198 @@
+  confirmed by the server since module startup
+  * `messages_failed` - number of messages to a RabbitMQ server that were
+  rejected by the server since module startup
+  * `message_publish_time` - amount of time it takes to publish a message to
```
Lacks information whether time is measured for all publishes or only successful ones.

fenek commented on test/mongoose_rabbit_worker_SUITE.erl:
```diff
@@ -0,0 +1,205 @@
+random_pid() ->
+    spawn(fun() -> ok end).
+
+lock_fun() ->
```
A bit of an overkill. Absolutely no gain from extracting it into separate function.

mkacper replied:
AFAIR I've put this funs into separate functions for the sake of tracing (I had some troubles when I was developing this suite). I can trace functions from the suite that wrap funs. Couldn't it be helpful? What do you reckon?

fenek replied:
OK, I accept such explanation for this and other functions. :) I think it would be nice to have a comment here like "BIFs extracted for tracing"

fenek commented on test/mongoose_rabbit_worker_SUITE.erl:
```diff
@@ -0,0 +1,205 @@
+        R -> unlocked
+    end.
+
+send_back_fun() ->
```
Overkill again.

fenek commented on test/mongoose_rabbit_worker_SUITE.erl:
```diff
@@ -0,0 +1,205 @@
+send_back_fun() ->
+    fun erlang:send/2.
+
+exception_fun() ->
```
And again.

mkacper commented on big_tests/tests/mod_event_pusher_rabbit_SUITE.erl:
```diff
@@ -0,0 +1,709 @@
+    rpc(mim(), gen_mod, stop_module, [Host, mod_event_pusher_rabbit]).
+
+start_rabbit_wpool(Host) ->
+    rpc(mim(), mongoose_wpool, ensure_started, []),
```
I've run the suite without calling this function and it didn't pass. It seems the pool is not started automatically.

fenek replied:
OK, I believe you, however it sounds strange. Looks like some kind of bug to investigate on other occasion.

mkacper commented on big_tests/tests/mod_event_pusher_rabbit_SUITE.erl:
```diff
@@ -0,0 +1,709 @@
+muc_host() ->
+    ?MUC_HOST.
+
+stanza_muc_enter_room(Room, Nick) ->
```
To be honest I just reused these helpers from some other SUITE :) Could you please tell me which of them can be replaced?

mkacper commented on doc/advanced-configuration/outgoing-connections.md:
```diff
@@ -9,6 +9,7 @@ The interface for outgoing connections management was unified and is now availab
 * `http` - pool of connections to various HTTP(S) servers MongooseIM can talk to, in example HTTP authentication backend or HTTP notifications
 * `elastic` - pool of connections to ElasticSearch server
 * `rdbms` - pool of connections to an RDBMS database
+* `rabbit` - pool of connections to a RabbitMQ server
```
I wouldn't say so without testing. The solution is tailored to RabbitMQ.

mkacper commented on doc/modules/mod_event_pusher_rabbit.md:
```diff
@@ -0,0 +1,198 @@
+### Worker selection strategy
+
+The module uses `mongoose_wpool` (which in turn uses `wpool` underneath) for
+managing worker processes  and [`best_worker` strategy](https://github.com/inaka/worker_pool#best_worker),
```
So maybe I'll just delete the link here? It's also placed in `mongoose_wpool` documentation. There is a link to this documentation somewhere above.

mongoose-im commented:
[6114.1](https://travis-ci.org/esl/MongooseIM/jobs/482971664) / Erlang 19.3 / small_tests / a190600754d415fd013c86e3f363f372a0297e87
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2127/6114/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2127/6114/small_tests.19.3/small/ct_run.mongooseim@localhost.2019-01-22_16.39.38/lib.mongooseim.logs/run.2019-01-22_16.39.39/suite.log.html)

mongoose-im commented:
[6117.1](https://travis-ci.org/esl/MongooseIM/jobs/483316856) / Erlang 19.3 / small_tests / b90cffb6bd1851a9f2623790a04ddb76d54e202a
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2127/6117/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2127/6117/small_tests.19.3/small/ct_run.mongooseim@localhost.2019-01-23_09.48.54/lib.mongooseim.logs/run.2019-01-23_09.48.55/suite.log.html)

mongoose-im commented:
[6118.1](https://travis-ci.org/esl/MongooseIM/jobs/483421154) / Erlang 19.3 / small_tests / a45ce171c35c16be2a42936bc8f55d99a2129b02
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2127/6118/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2127/6118/small_tests.19.3/small/ct_run.mongooseim@localhost.2019-01-23_14.23.19/lib.mongooseim.logs/run.2019-01-23_14.23.20/suite.log.html)
